In [80]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [85]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
file_path = "/Users/LENOVO/Library/CloudStorage/OneDrive-Personal/Bootcamp/Module 20/credit-risk-classification/Starter_Code/Resources/lending_data.csv"

# Load the data into a Pandas DataFrame
df_lending_data = pd.read_csv(file_path)

# Review the DataFrame
df_lending_data.head()


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [88]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending_data["loan_status"]


# Separate the X variable, the features
X = df_lending_data.drop(columns=["loan_status"])

In [90]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [92]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [95]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [111]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [114]:
# Make a prediction using the testing data
testing_predictions = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [123]:
# Generate a confusion matrix for the model
conf_matrix = confusion_matrix(y_test, testing_predictions)
print(conf_matrix)

[[18652    99]
 [   65   568]]


In [127]:
from sklearn.metrics import classification_report

# Generate the classification report
class_report = classification_report(y_test, testing_predictions)

# Print the classification report
print(class_report)


              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18751
           1       0.85      0.90      0.87       633

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.93     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** **Healthy Loans (Label: 0)**
**Precision:** 1.00. This indicates that out of all the loans predicted as healthy, 100% were healthy.

**Recall:** 0.99- This means that the model correctly identified 99% of all actual healthy loans.

**F1-Score:** 1.00.- This is a harmonic mean of precision and recall, showing a balanced measure of the model's accuracy for healthy loans.

**High-Risk Loans (Label: 1)**
**Precision:** 0.85- This indicates that out of all the loans predicted as high-risk, 85% were high-risk.

**Recall:** 0.90 This means that the model correctly identified 90% of all actual high-risk loans.

**F1-Score:** 0.87. -This represents a balanced measure of the model's accuracy for high-risk loans, considering both precision and recall.

**Overall Performance**
**Accuracy: 0.99**

.The model's overall accuracy is 99%, meaning it correctly classified 99% of all loans.

**Macro Average:**

.Precision: 0.92, Recall: 0.95, F1-Score: 0.93
.These are the average metrics calculated without considering class imbalance.

**Weighted Average:**

**Precision:** 0.99, Recall: 0.99, F1-Score: 0.99
These metrics consider the support (number of instances) of each class, giving more weight to the majority class.

**Interpretation**
.The model performs exceptionally well in predicting healthy loans with perfect precision and very high recall.

.For high-risk loans, while the precision is slightly lower at 85%, the recall is high at 90%, indicating that the model is good at identifying most of the high-risk loans.

.The overall high accuracy and strong F1-scores suggest that the model is robust and reliable for credit risk analysis.

---